share/doc/wake/wake-for-fsharp-developers.md

```diff
@@ -0,0 +1,150 @@
+# Wake for F# developers
+
+F# and Wake are both ML-dialects, so they share very similar heritages as far as programming languages go. The major difference is that F# is intended to be a general-purpose programming language and is multi-paradigm since it also supports imperative and OOP programming, whereas Wake is intended to be a programming language to support a build system and does not support the imperative or OOP programming paradigms. However, while simple, the Wake programming language is rather powerful.
+
+In many cases, the difference between F# and Wake are syntactical only. Below is a collection of comparisons to ease the transition between the two.
+
+## Definitions
+
+In F#, bindings are created with `let` and recursive functions are created with `let rec`.
+
+```fsharp
+let numberTwo = 2
+
+let rec fibonacci (n: int) : int =
+    match n with
+    | 0 | 1 -> n
+    | n -> fibonacci (n-1) + fibonacci (n-2)
+```
+
+In Wake, bindings are created with `def` and do not require any keyword if they are recursive.
+
+```wake
+def numberTwo = 2
+
+def fibonacci (n: Integer): Integer =
+    match n
+        0 = n
+        1 = n
+        n = fibonacci (n-1) + fibonacci (n-2)
+```
+
+## Lists
+
+In F#, cons is `::` and the empty list is `[]`. There is literal list syntax with `[]` and `;` as a separator.
+
+```fsharp
+> 1 :: 2 :: 3 :: [];;
+val it: int list = [1; 2; 3]
+
+> [1; 2; 3];;
+val it: int list = [1; 2; 3]
+```
+
+In Wake, cons is [`,`](https://github.com/sifive/wake/blob/master/share/wake/lib/core/list.wake#L43-L50) and the empty list is [`Nil`](https://github.com/sifive/wake/blob/master/share/wake/lib/core/list.wake#L41-L42). There is currently no literal list syntax in Wake. You often need to surround a list value with parentheses, such as `(1, 2, 3, Nil)`.
+
+```wake
+> wake -vx '1, 2, 3, Nil'
+1, 2, 3, Nil: List Integer = 1, 2, 3, Nil
+```
+
+## Discriminated unions vs `data` types
+
+F#
+```fsharp
+type Shape =
+    | Circle of radius: float
+    | Square of side: float
+    | Rectangle of length: float * width: float
+```
+
+Wake
+```wake
+export data Shape =
+    Circle (radius: Double)
+    Square (side: Double)
+    Rectangle (length: Double) (width: Double)
+```
+
+## Pattern Matching
+
+F#
+```fsharp
+let rec area shape =
+    match shape with
+    | Circle radius             -> radius * radius
+    | Square side               -> area (Rectangle (side, side))
+    | Rectangle (length, width) -> length * width
+```
+
+Wake
+```wake
+export def area (shape: Shape) =
+    match shape
+        Circle radius          = pi *. (radius ^. 2.0)
+        Square side            = area (Rectangle side side)
+        Rectangle length width = length *. width
+```
+
+Note that in F#, recursive functions require the `rec` keyword, whereas in Wake, recursive functions require no special keyword.
+
```

Comment:
I'm guessing this got left behind after reorganizing the file at some point?
```suggestion
```

Reply:
That's exactly what happened. [Removed here.](https://github.com/sifive/wake/pull/1119/commits/fdbe702140ce35f66cd39a064b2f5ae439c1a33f) (Needed to remove an extra newline.)

Reply:
I actually had the newline removed in the suggestion as well, but apparently GitHub doesn't like line+blankline -> (remove both) changes in comments.

Reply:
Ah, interesting. No worries!